spray-json/shared/src/main/scala/spray/json/JsonParser.scala

```diff
@@ -165,51 +170,148 @@ class JsonParser(input: ParserInput, settings: JsonParserSettings = JsonParserSe
+    }*/
+
+    /*
+     * Do full string parsing. This subsumes the functionality from parseOneFast, however, just using
+     * this one for all parsing is significantly slower (> 10 %). A possible reason might be that using
+     * this function for all string parsing has worse branch prediction than the simple and fast variant?
+     */
+    @tailrec def parseOneSlow(cursor: Int, charsRead: Int): String = {
+      val b = input.charAt(cursor)
+      charBuffer(charsRead) = b
+
+      if (b == '"') {
+        input.setCursor(cursor) // now at end '"'
+        advance()
+        new String(charBuffer, 0, charsRead)
+      } else if (b == '\\') {
+        val skip = escaped(cursor, charsRead)
+        parseOneSlow(cursor + skip, charsRead + 1)
+      } else if (input.needsDecoding && b >= 128) {
+        val res = unicodeChar(b, cursor, charsRead)
+        val skip = res & 0x7
+        val read = res >> 4
+        parseOneSlow(cursor + skip, charsRead + read + 1)
+      } else
+        parseOneSlow(cursor + 1, charsRead + 1)
     }
 
-  private def `escaped`() = {
-    def hexValue(c: Char): Int =
-      if ('0' <= c && c <= '9') c - '0'
-      else if ('a' <= c && c <= 'f') c - 87
-      else if ('A' <= c && c <= 'F') c - 55
-      else fail("hex digit")
-    def unicode() = {
-      var value = hexValue(cursorChar)
-      advance()
-      value = (value << 4) + hexValue(cursorChar)
-      advance()
-      value = (value << 4) + hexValue(cursorChar)
-      advance()
-      value = (value << 4) + hexValue(cursorChar)
-      appendSB(value.toChar)
+    def escaped(cursor: Int, charsRead: Int): Int =
+      input.byteAt(cursor + 1) match {
+        case c @ ('"' | '/' | '\\') =>
+          charBuffer(charsRead) = c.toChar
+          2
+        case 'b' =>
+          charBuffer(charsRead) = '\b'
+          2
+        case 'f' =>
+          charBuffer(charsRead) = '\f'
+          2
+        case 'n' =>
+          charBuffer(charsRead) = '\n'
+          2
+        case 'r' =>
+          charBuffer(charsRead) = '\r'
+          2
+        case 't' =>
+          charBuffer(charsRead) = '\t'
+          2
+        case 'u' =>
+          def hexValue(c: Byte): Int =
+            if ('0' <= c && c <= '9') c - '0'
+            else if ('a' <= c && c <= 'f') c - 87
+            else if ('A' <= c && c <= 'F') c - 55
+            else fail("hex digit")
+
+          charBuffer(charsRead) =
+            ((hexValue(input.byteAt(cursor + 2)) << 12) |
+              (hexValue(input.byteAt(cursor + 3)) << 8) |
+              (hexValue(input.byteAt(cursor + 4)) << 4) |
+              hexValue(input.byteAt(cursor + 5))).toChar
+
+          6
+        case _ => fail("JSON escape sequence")
+      }
+    def unicodeChar(b: Char, cursor: Int, charsRead: Int): Int = {
+      if ((b & 0xE0) == 0xC0) { // two byte sequence
+        charBuffer(charsRead) = (((b & 0x1f) << 6) |
+          (input.byteAt(cursor + 1) & 0x3f)).toChar
```

Comment:
here and bellow all subsequent bytes after leading one should be checked if they are well formed (have `10` highest bits)